java/src/main/java/ai/rapids/cudf/MemoryListener.java

```diff
@@ -0,0 +1,137 @@
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ */
+
+package ai.rapids.cudf;
+
+import java.util.HashSet;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * Provides a callback API that can be used to track device memory usage.  The
+ * reported usage is neither exact nor complete, but it should give a picture of memory usage by
+ * this library. It does not currently track internal intermediate allocations done by the library.
+ * It also does not track any memory fragmentation that might prevent it from being used in the
+ * future.
+ * <br />
+ * The callbacks take the form of <ol>
+ *   <li>prediction</li>
+ *   <li>allocation</li>
+ *   <li>endPrediction</li>
+ *   <li>deallocation</li>
+ * </ol>
+ * An allocation indicates that memory was allocated. One or more allocation call should typically
+ * happen between a prediction and endPrediction call, but it is not guaranteed. A prediction
+ * indicates an educated guess as to how much memory an operation is likely to take. In some cases
+ * we know exactly what it takes, but in others it is really just a guess and it may guess too much
+ * or too little memory. Typically this is used to try and tell if we are getting close to memory
+ * limits and try to reduce the amount of memory being used.  In those cases a prediction should be
+ * treated like a memory reservation that is deducted from when an allocation happens until it is
+ * used up.
+ * <br />
+ * Please note that if your code is threaded the MemoryListener will be called from multiple
+ * threads and each prediction must be treated in a thread local way.
+ */
+public abstract class MemoryListener {
+  /**
+   * A prediction about how much memory is about to be used for an operation.
+   * @param amount the number of bytes predicted to be used.
+   * @param note a string that can be used for debugging to help keep track of what is making the
+   *             prediction.
+   */
+  public abstract void prediction(long amount, String note);
+
+  /**
+   * An actual memory allocation has occurred.
+   * @param amount the number of bytes actually allocated.
+   * @param id a unique number that can be used for debugging. The same number will be sent to
+   *           deallocate.
+   */
+  public abstract void allocation(long amount, long id);
+
+  /**
+   * Indicates that the previous prediction on this thread is no longer in force. Predictions should
+   * never nest.  Only one prediction should ever be in force at a time.
```

Comment:
I thought it would be up to the end user to decide what they want to do with the information.  They can keep track of the exact memory that has been allocated and ignore the predictions, or they can use the predictions to make a reservation and then deduct memory from that reservation as it is allocated.  It all depends on what you want to do.